lib/data/repository/default_marvell_repository.dart

```diff
@@ -0,0 +1,113 @@
+    _repository.remoteDataSource = remoteDataSource;
+    return _repository;
+  }
+
+  @override
+  Future<Character> getCharacterDetail(String characterId) async {
+    try {
+      var httpResponse = await remoteDataSource.getCharacterDetail(characterId);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+        Character? tempCharacter = data?.results?.map((item) => item.fromApiToCharacter("standard_xlarge")).single;
+        Character character = tempCharacter as Character;
+        return character;
+      } else {
+        throw DataRetrieveException();
+      }
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  @override
+  Future<List<Character>> getCharacters([int offset = 0]) async {
+    if (offset == 0) {
+      List<Character> characters = List.empty();
+      try {
+        characters = await _getRemote(offset);
+      } on Exception {
+        characters = await _getLocal();
+      }
+      return characters;
+    }
+    try {
+      return _getRemote(offset);
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  Future<List<Character>> _getRemote(int offset) async {
+    List<Character> characters = List.empty();
+    try {
+      var httpResponse = await remoteDataSource.getCharacters(offset);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+        List<Character>? tempCharacters =
+            data?.results?.map((item) => item.fromApiToCharacter("standard_medium")).toList();
+        characters = tempCharacters ?? characters;
+        await _saveCharactersInDatabase(characters);
+      }
+      return characters;
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  Future<List<Character>> _getLocal() async {
+    try {
+      return await localDataSource.characterDao.getAllCharacters();
+    } catch (error) {
+      throw DataRetrieveException();
+    }
+  }
+
+  @override
+  Future<List<Series>> getSeries(String characterId) async {
+    try {
+      var httpResponse = await remoteDataSource.getSeries(characterId);
+      if (httpResponse.response.statusCode == 200) {
+        var data = httpResponse.data.data;
+
+        List<Series>? tempSeries = data?.results?.map((item) => item.toSeries("portrait_medium")).toList();
+        List<Series> series = tempSeries ?? List.empty();
+        return series;
+      } else {
+        throw DataRetrieveException();
+      }
+    } on DioError catch (error) {
+      throw _getException(error);
+    }
+  }
+
+  Future<void> _saveCharactersInDatabase(List<Character> characters) async {
+    await localDataSource.characterDao.deleteOldData();
+    await localDataSource.characterDao.insertData(characters);
+  }
+
+  Exception _getException(DioError error) {
+    Exception exception = NoConnectionException();
+    if (error.type == DioErrorType.connectTimeout) {
```

Comment:
Cannot you get `DioErrorType.connectTimeout` if there is no connection? 🤔

Reply:
if error type "other" in case when not connection, we cannot get connectTimeout.
and code snippet above only for dio error type "other" or other types from errors enum.